crates/orderbook/src/orderbook.rs

```diff
@@ -449,63 +452,77 @@ impl Orderbook {
+                    let executed_amounts = solution.orders.iter().find_map(|o| match o {
+                        solver_competition::Order::Legacy { .. } => None,
+                        solver_competition::Order::Colocated {
+                            id,
+                            sell_amount,
+                            buy_amount,
+                        } => (id == uid).then_some(dto::order::ExecutedAmounts {
+                            sell: *sell_amount,
+                            buy: *buy_amount,
+                        }),
+                    });
+                    dto::order::SolutionInclusion {
+                        solver: solution.solver,
+                        executed_amounts,
                     }
-                    OrderEventLabel::Traded => dto::order::Status::Traded(fetch_solutions().await?),
-                    OrderEventLabel::Cancelled => dto::order::Status::Cancelled,
-                    OrderEventLabel::Filtered => dto::order::Status::Open,
-                    OrderEventLabel::Invalid => dto::order::Status::Open,
-                };
+                })
+                .collect::<Vec<_>>()
+        };
+
+        let latest_competition = async {
+            let competition = self.database.load_latest_competition().await?;
+            Ok::<_, anyhow::Error>(solutions(competition))
+        };
+
+        // Once an order was executed we always want to return `Traded` with the
+        // competition data of the **first** time it was traded for a stable result.
+        // Under some circumstances it can happen that the latest state of an already
+        // executed order is not `Traded`. To detect that we first check the trades
+        // table and return the appropriate competition data.
+        let trades = self
+            .database
+            .trades(&TradeFilter {
+                owner: None,
+                order_uid: Some(*uid),
+            })
+            .await?;
 
-                Ok(Some(status))
+        match trades.first().map(|trade| trade.tx_hash) {
+            Some(Some(tx_hash)) => {
+                let competition = self
+                    .database
+                    .load_competition(Identifier::Transaction(tx_hash))
+                    .await?;
+                return Ok(Some(dto::order::Status::Traded(solutions(competition))));
             }
+            // order executed but not fully indexed and processed
+            Some(None) => {
+                return Ok(Some(dto::order::Status::Traded(latest_competition.await?)));
+            }
+            None => (),
         }
+
+        let latest_event = self.database.latest_order_event(uid).await?;
+        let status = match latest_event.context("no event")?.label {
+            OrderEventLabel::Ready => dto::order::Status::Active,
+            OrderEventLabel::Created => dto::order::Status::Scheduled,
+            OrderEventLabel::Considered => dto::order::Status::Solved(latest_competition.await?),
```

Comment:
not related to this fix but aren't `Considered` and `Executing` supposed to be saved only once the latest competition is saved to db, so that we make sure that they do not return some other competition?

Reply:
Good point. We could move storing those into the code that stores the competition itself as we can easily recover that data from the competition.
Should be done in a follow up PR, though.